### ISSUE TITLE
Explore multithreading/multiprocessing

### ISSUE THREAD

**Issue by thebigmunch** [thebigmunch/google-music-scripts#19]
I once experimented with using the multiprocessing module with gmusicapi-scripts. There was some kind of bug/interaction that prevented this from working properly back then.

I'd like to take another look at using multiple threads or multiple processes to improve the uploading/downloading experience. I'd look at using ``ThreadPoolExecutor``/``ProcessPoolExecutor`` from [concurrent.futures](https://docs.python.org/3/library/concurrent.futures.html).

I'm considering adding options for both, as there are cases where just using threads might be better (sample generation disabled and uploading MP3s) and cases where using processes might be better (generating samples and transcoding non-MP3s). These options would be mutually exclusive.

**Comment by jbaum98** [comment]
I think you can actually use `asyncio` or some other event loop with `async/await` to take advantage of parallelism for both IO-bound work and CPU-bound work. The event loop itself handles waiting for IO-bound tasks, like a `ThreadPoolExecutor` would, and to get around the GIL you can run CPU-bound tasks on a `ProcessPoolExecutor`.

I have an example [here](https://gist.github.com/jbaum98/51baa880b9108aef23ffd5edcaca56ec) where I basically sleep in a way that the event loop can handle for 2 seconds and then sleep in a way that blocks it for 1 second, to simulate IO-bound and CPU-bound tasks respectively.

In the first one, I just do the CPU-bound task with handling it specially.
```python
async def task(loop, executor, async_sleep, sync_sleep):
    times = [time.monotonic()]
    # This is like doing some IO
    await asyncio.sleep(async_sleep)
    times.append(time.monotonic())
    # This is like doing a lot of computation
    time.sleep(sync_sleep)
    times.append(time.monotonic())
    return times
```
The blue parts are IO-bound, and they are parallelized, but once an orange part starts the other blue orange parts have to wait

![sync](https://user-images.githubusercontent.com/5283991/51196806-e3b9c900-18bd-11e9-92b1-43ba177bfd8b.png)

Here, I execute the CPU bound tasks on a different process, and even the CPU-bound parts are parallelized.
```python
async def task(loop, executor, async_sleep, sync_sleep):
    times = [time.monotonic()]
    # This is like doing some IO
    await asyncio.sleep(async_sleep)
    times.append(time.monotonic())
    # This is like doing a lot of computation
    await loop.run_in_executor(executor, time.sleep, sync_sleep)
    times.append(time.monotonic())
    return times
```
![async](https://user-images.githubusercontent.com/5283991/51196911-1cf23900-18be-11e9-8fb9-e40e8cac42e7.png)

**Comment by thebigmunch** [comment]
Yeah, async in Python is too much of a pain right now. And asyncio's API and docs were so horrible, they've had to make some substantial changes/additions that still don't make it particularly nice yet. And that's if you want to go 3.7+, which I don't at this point. Those changes are still provisional, so they could be changed/removed. E.g. ``asyncio.run`` is one of those provisional changes that I'd have to go 3.7+ to use. Seems trivial, but there's a reason it was added. So I'd use Trio over it if I were to go that route anyway. And the big thing is that so little of the code actually needs to go concurrent (just the uploading/downloading).

But, this is something so small and simple that a ``ThreadPoolExecutor`` in the uploading function is the way to go. I was playing with it the other day and works great without having to muck about with much code. ``ProcessPoolExecutor`` on the other hand wasn't working properly, and I didn't bother to check it. Fact is, even the CPU stuff isn't going to be intensive or long-lived, so I may end up not even bothering to try to suss out the problem. Most of the benefit is had with threads rather than processes anyway.

PS I've been dying to do things with async in Python for years now. That's part of the reason I wrote a [sans-io](https://github.com/brettcannon/sans-io) Google Music wrapper ([google-music-proto](https://github.com/thebigmunch/google-music-proto)) as the base of everything. Some day...
